core/server/common/src/main/java/alluxio/master/journal/JournalSystem.java

```diff
@@ -0,0 +1,184 @@
+   * entries. However, no entries may be written until the journal system has been started, and
+   * entries may only be written when the journal system is in PRIMARY mode.
+   *
+   * When the journal is started in secondary mode, it will call
+   * {@link JournalEntryStateMachine#processJournalEntry(JournalEntry)} and
+   * {@link JournalEntryStateMachine#resetState()} to keep the state machine's state in sync with
+   * the entries written to the journal.
+   *
+   * Although the JournalSystem interface is generally not threadsafe, it is required that
+   * implementations of JournalSystem support concurrent invocations of this method.
+   *
+   * @param stateMachine the state machine to create the journal for
+   * @return a new instance of {@link Journal}
+   */
+  Journal create(JournalEntryStateMachine stateMachine);
+
+  /**
+   * Starts the journal system.
+   *
+   * All journals must be created before starting the journal system. This method will block until
+   * the journal system is successfully started. The journal always starts in secondary mode.
+   */
+  void start() throws InterruptedException, IOException;
+
+  /**
+   * Stops the journal system.
+   */
+  void stop() throws InterruptedException, IOException;
+
+  /**
+   * Sets the mode of the journal to either primary or secondary. This method requires that the
+   * journal system has been started.
+   *
+   * When journal entries are applied, the journal will only update the state machine if in
+   * secondary mode. In primary mode, it is expected that the state machine will update its own
+   * state before it writes entries to the journal.
+   *
+   * @param mode the mode to set
+   */
+  void setMode(Mode mode);
+
+  /**
+   * Formats the journal system.
+   */
+  void format() throws IOException;
+
+  /**
+   * @return whether the journal system has been formatted
+   */
+  boolean isFormatted() throws IOException;
+
+  /**
+   * Builder for constructing a journal system.
+   */
+  class Builder {
+    private URI mLocation;
+    private long mQuietTimeMs =
+        Configuration.getMs(PropertyKey.MASTER_JOURNAL_TAILER_SHUTDOWN_QUIET_WAIT_TIME_MS);
+
+    /**
+     * Creates a new journal system builder.
+     */
+    public Builder() {}
+
+    /**
+     * @param location the location for the journal system
+     * @return the updated builder
+     */
+    public Builder setLocation(URI location) {
+      mLocation = location;
+      return this;
+    }
+    /**
+     * @param quietTimeMs before upgrading from SECONDARY to PRIMARY mode, the journal will wait
+     *        until this duration has passed without any journal entries being written.
+     * @return the updated builder
+     */
+    public Builder setQuietTimeMs(long quietTimeMs) {
+      mQuietTimeMs = quietTimeMs;
+      return this;
+    }
+
+    /**
+     * @return a journal system
+     */
+    public JournalSystem build() {
+      return new UfsJournalSystem(mLocation, mQuietTimeMs);
```

Comment:
oh interesting idea exposing it to users. it could be useful to measure journaling impact on performance. I'll make it a configuration option